test/Jamfile.v2

```diff
@@ -7,10 +7,18 @@
 project libs/uuid/test
     : requirements
 
-      # boost::random requires this setting for a warning free build:
+	  # define BOOST_UUID_SOURCE to defeat the default BOOST_ALL_NO_LIB in boost.jam
+	  <define>BOOST_UUID_SOURCE=1
+
+      # BOOST_AUTO_LINK does not work outside of MSVC on Windows so this makes cygwin, mingw-w64 work:
+      <toolset>gcc,<target-os>windows:<linkflags>-lbcrypt 
```

Comment:
We could link here regardless of whether the toolset is gcc (there's also clang btw), and this will fix the lack of autolink when using Boost.Build. But the problem is that we don't necessarily know what to link to.

`<linkflags>-lbcrypt` isn't correct though, I think that we need `lib bcrypt ;` and then `<library>bcrypt` here.

Reply:
I will try that and see what it does to the mocks.

Reply:
Is there any way to negate this (`<target-os>windows:<library>bcrypt`), so that I can set it in project requirements so everything links against it by default, but then in a specific mock based test negate it so it does NOT link against bcrypt?

Reply:
I vaguely recall that one could use `<x>y:-<library>bcrypt` to remove it, but I won't be surprised if it doesn't work and I'm imagining things. :-)

Reply:
Turns out I'm not, in this case. http://www.boost.org/build/doc/html/bbv2/overview/targets.html#bbv2.overview.targets.requirements

Reply:
A change as proposed would break, for example, MinGW (32-bit) where bcrypt is not part of the distribution, so the bjam directives would need to be as smart as the preprocessor selection logic.  In fact I bet mingw is already broken, but it hasn't been updated since 2013 so I'm not going to worry about it.  I think I'll keep the auto-link and global gcc/windows directives as-is as they work in CI (well, I'll change it from linkflags to library).

Reply:
Not sure what change would break what, as you're already linking to bcrypt in the gcc case.

Reply:
> In fact I bet mingw is already broken

There are many mingws, so some could be. I tried the one I have here, and it passes the tests except for test_serialization.exe, which crashes. But that may well be Serialization's fault.

Reply:
Appveyor has the plain mingw in C:\mingw, so you could test it in addition to mingw-w64, which you currently have. See f.ex. https://ci.appveyor.com/project/pdimov/system

Reply:
Take a look at my appveyor config; I found that a little while ago and marked some jobs as allowed to fail.  Serialization has a large bug backlog (40 to 50) in trac... I don't know if I reported it.

Reply:
Odd, test_serialization passed the second time. Mystery.

Reply:
Instead of not testing anything, it might be better to disable just test_serialization on mingw and keep testing the rest.

Reply:
Cygwin, on the other hand, passes with flying colors.

Reply:
I am unable to use the syntax: `<target-os>windows:-<library>bcrypt`, or `<target-os>windows:-<define>BOOST_ALL_NO_LIB` - both fail in bjam as an invalid target-os.  I'm going back to the logic I had.

Reply:
If only there existed a simple function in `advapi32` that we could just call and avoid all that hassle.